lib/pushmi_pullyu/aip/downloader.rb

```diff
@@ -15,248 +16,198 @@ class PushmiPullyu::AIP::Downloader
+    # proper authentication mechanism is setup on jupiter
+    # https://github.com/ualbertalib/jupiter/pull/1370#issuecomment-561799351
+    request.basic_auth('admin', 'admin@ualberta.ca')
+
+    Net::HTTP.start(uri.hostname, uri.port) do |http|
+      http.request(request)
+    end
+  end
+
+  def download_and_log(path_spec)
+    output_file = path_spec[:local]
+    remote = path_spec[:remote]
+    log_downloading(remote, output_file)
 
-    log_fetching(fedora_fetcher.object_url(path_spec.remote), output_file)
+    response = authenticate_and_request(remote)
 
-    is_rdf = (output_file =~ /\.n3$/)
-    should_add_user_email = path_spec.to_h.fetch(:should_add_user_email, false)
+    is_success = false
+    if response.is_a?(Net::HTTPSuccess)
+      file = File.open(output_file, 'wb')
+      file.write(response.body)
+      file.close
+      # Response was a success and the file was saved to output_file
+      is_success = File.exist? output_file
+    end
 
-    is_success = fedora_fetcher.download_object(output_file,
-                                                url_extra: path_spec.remote,
-                                                optional: path_spec.optional,
-                                                is_rdf: is_rdf,
-                                                should_add_user_email: should_add_user_email)
     log_saved(is_success, output_file)
+    raise JupiterDownloadError unless is_success
+  end
+
+  def get_file_paths(url)
+    response = authenticate_and_request(url)
+    JSON.parse(response.body, symbolize_names: true)
   end
 
-  def create_and_log_file_order_list(url)
-    output_file = object_aip_paths.file_ordering.local
-    PushmiPullyu::Logging.log_aip_activity(@aip_directory,
-                                           "#{@noid}: #{output_file} -- creating from #{url} ...")
-    PushmiPullyu::AIP::FileListCreator.new(url, output_file, member_file_set_uuids).run
-    PushmiPullyu::Logging.log_aip_activity(@aip_directory,
-                                           "#{@noid}: #{output_file} -- created")
+  def object_uri
+    aip_api_url = PushmiPullyu.options[:jupiter][:aip_api_url]
+    @object_uri ||= "#{aip_api_url}/#{@entity[:type]}/#{@entity[:uuid]}"
   end
 
   ### Logging
 
-  def log_fetching(url, output_file)
-    message = "#{@noid}: #{output_file} -- fetching from #{url} ..."
+  def log_downloading(url, output_file)
+    message = "#{@entity_identifier}: #{output_file} -- Downloading from #{url} ..."
     PushmiPullyu::Logging.log_aip_activity(@aip_directory, message)
   end
 
   def log_saved(is_success, output_file)
-    message = "#{@noid}: #{output_file} -- #{is_success ? 'saved' : 'not_found'}"
+    message = "#{@entity_identifier}: #{output_file} -- #{is_success ? 'Saved' : 'Failed'}"
     PushmiPullyu::Logging.log_aip_activity(@aip_directory, message)
   end
 
   ### Directories
 
   def aip_dirs
-    @aip_dirs ||= OpenStruct.new(
+    @aip_dirs ||= {
```

Comment:
I have two reasons why I moved away from OpenStructs:

1. To my understanding OpenStructs are slow because they are just a wrapper around hashes.
2. In some cases the OpenStructs were not using the syntactical sugar they provide, that is, the OpenStruct variables were called as from square bracket operator instead of dot operator ( [:my_variable] vs .myvariable )

I know speed is not a big deal here but since OpenSctructs were only used in this one file, and without the syntactic sugar, I did not see a problem to switch to hashes.

Structs are faster than hashes, but they seemed like an overkill.

Reply:
I'm not married to Structs or anything. I do wonder if this has all gotten complicated enough that some small classes would make this all a bit cleaner, than our previous Structs of Structs or Hash of hashes approach, though.

Reply:
I see what you mean. In the past when I've dealt with more complex structures in hash format I found it was better to flatten the hash out. This worked to improve readability and performance. This would look something like this:
```
@object_aip_paths ||= {
      main_object_remote: object_uri,
      main_object_local: "#{aip_dirs[:metadata]}/object_metadata.n3",  
      file_sets_remote: "#{object_uri}/filesets",
      file_sets_local: "#{aip_dirs[:files_metadata]}/file_order.xml",
      # This is downloaded for processing but not saved
      file_paths_remote: "#{object_uri}/file_paths",    
    }.freeze
```

Reply:
I like this idea 👍 . I think it's the _nesting_  -- where we pass a smaller hash out of a hash that's pulled out of another hash -- so complicated.